File: Simulator/Components/Printer.vb
Imports System.Text

Public Class Printer

    Public Property Receipt As Models.Receipt

    Public Sub Print()
        Me.PanelStatus.BackColor = Color.Orange
        Me.TimerPrinting.Start()
    End Sub

    Public Sub Discart()
        Me.PanelStatus.BackColor = IIf(Me.Enabled, Color.LimeGreen, Color.Red)
        Me.Receipt = Nothing
        Me.PanelPaper.Visible = False
    End Sub

    Private Sub TimerDiscart_Tick(sender As Object, e As EventArgs) Handles TimerDiscart.Tick
        Me.Discart()
        Me.TimerDiscart.Stop()
    End Sub

    Private Sub TimerPrinting_Tick(sender As Object, e As EventArgs) Handles TimerPrinting.Tick
        Me.PanelPaper.Visible = True
        Me.TimerPrinting.Stop()
        Me.TimerDiscart.Start()
    End Sub

    Private Sub Paper_Click(sender As Object, e As EventArgs) Handles PanelPaper.Click, LabelPaper.Click
        Dim _ReceiptPreview As New Form
        With _ReceiptPreview

            .FormBorderStyle = FormBorderStyle.FixedSingle
            .StartPosition = FormStartPosition.CenterScreen
            .Icon = Me.FindForm.Icon
            .Padding = New Padding(3)
            .Size = New Size(240, 320)
            .MaximizeBox = False

            Dim _receipt As New Panel
            _receipt.Padding = New Padding(6)
            _receipt.BackColor = Color.White
            _receipt.Font = New Font("Consolas", 8)
            _receipt.BorderStyle = BorderStyle.FixedSingle
            _receipt.Dock = DockStyle.Fill
            _receipt.Controls.Add(New Label() With {.Dock = DockStyle.Fill, .AutoSize = False, .Text = Generate(Me.Receipt)})

            .Controls.Add(_receipt)

            .Show()

        End With
        Me.Discart()
    End Sub

    Private Function SecureIt(source As String, Optional chr As Char = "+", Optional visible As Integer = 4) As String
        If source.Length > visible Then
            Dim hiddend = source.Substring(0, source.Length - visible).Length
            Return StrDup(hiddend, chr) & source.Substring(hiddend)
        Else
            Return source
        End If
    End Function


    Public Function Generate(source As Models.Receipt) As String
        Dim _template As String = My.Resources.Receipt

        Dim _result As String = _template _
            .Replace("{BANK}", source.Bank) _
            .Replace("{ATM}", source.ATM) _
            .Replace("{CODE}", source.Code) _
            .Replace("{DATE}", source.Date.ToShortDateString) _
            .Replace("{TIME}", source.Date.ToShortTimeString) _
            .Replace("{CARD}", SecureIt(source.Card)) _
            .Replace("{OPERATION}", source.Operation) _
            .Replace("{ACCOUNT}", SecureIt(source.Account)) _
            .Replace("{DETAIL}", source.Detail)

        Return _result
    End Function

    Private Sub _EnabledChanged(sender As Object, e As EventArgs) Handles Me.EnabledChanged
        Me.PanelStatus.BackColor = IIf(Me.Enabled, Color.LimeGreen, Color.Red)
    End Sub

End Class

File: Simulator-CSharp/Components/Printer.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace ATM.Components
{
    public partial class Printer : UserControl
    {
        public Printer()
        {
            InitializeComponent();
        }

        public Models.Receipt Receipt { get; set; }

        public void Print()
        {
            panelStatus.BackColor = Color.Orange;
            timerPrinting.Start();
        }

        public void Discart()
        {
            panelStatus.BackColor = (Enabled ? Color.LimeGreen : Color.Red);
            Receipt = null;
            panelPaper.Visible = false;
        }

        private string SecureIt(string source, char chr = '+', int visible = 4)
        {
            if (source.Length > visible)
            {
                var hiddend = source.Substring(0, source.Length - visible).Length;
                return source.PadRight(hiddend, chr);
            }
            else {
                return source;
            }
        }

        public string Generate(Models.Receipt source)
        {
            string _template = Properties.Resources.Receipt;

            string _result = _template
                .Replace("{BANK}", source.Bank)
                .Replace("{ATM}", source.ATM)
                .Replace("{CODE}", source.Code)
                .Replace("{DATE}", source.Date.ToShortDateString())
                .Replace("{TIME}", source.Date.ToShortTimeString())
                .Replace("{CARD}", SecureIt(source.Card))
                .Replace("{OPERATION}", source.Operation)
                .Replace("{ACCOUNT}", SecureIt(source.Account))
                .Replace("{DETAIL}", source.Detail);

            return _result;
        }

        private void timerDiscart_Tick(object sender, EventArgs e)
        {
            Discart();
            timerDiscart.Stop();
        }

        private void timerPrinting_Tick(object sender, EventArgs e)
        {
            panelPaper.Visible = true;
            timerPrinting.Stop();
            timerDiscart.Start();
        }

        private void panelPaper_Click(object sender, EventArgs e)
        {
            Form _ReceiptPreview = new Form();


            {
                _ReceiptPreview.FormBorderStyle = FormBorderStyle.FixedSingle;
                _ReceiptPreview.StartPosition = FormStartPosition.CenterScreen;
                _ReceiptPreview.Icon = this.FindForm().Icon;
                _ReceiptPreview.Padding = new Padding(3);
                _ReceiptPreview.Size = new Size(240, 320);
                _ReceiptPreview.MaximizeBox = false;

                Panel _receipt = new Panel();
                _receipt.Padding = new Padding(6);
                _receipt.BackColor = Color.White;
                _receipt.Font = new Font("Consolas", 8);
                _receipt.BorderStyle = BorderStyle.FixedSingle;
                _receipt.Dock = DockStyle.Fill;
                _receipt.Controls.Add(new Label
                {
                    Dock = DockStyle.Fill,
                    AutoSize = false,
                    Text = Generate(this.Receipt)
                });

                _ReceiptPreview.Controls.Add(_receipt);

                _ReceiptPreview.Show();

            }
            this.Discart();
        }

        private void _EnabledChanged(object sender, EventArgs e)
        {
            panelStatus.BackColor = (Enabled ? Color.LimeGreen : Color.Red);
        }
    }
}
